doc/reference/trd-digest.md

```diff
@@ -0,0 +1,314 @@
+in RAM and passing it without the `mut` qualifier in a split-phase
+operation can discard its mutability 
+(see Rule 5 in [TRD2](./trd2-hil-design.md)). Digest supports immutable
+data because many services need to compute digests over large, read-only data
+in flash. One example of this is the kernel's process loader, which needs
+to check that process images are not corrupted. Because digests are 
+computationally inexpensive, copying the data from flash to RAM in order
+to compute a digest is a large overhead. Furthermore, the data input can
+be large (tens or hundreds of kilobytes). Therefore `DigestData` and 
+`ClientData` support both mutable and immutable inputs.
+
+Clients provide input to `DigestData` through the `LeasableBuffer`
+and `LeasableMutableBuffer` types. These allow a client to ask a
+digest engine to compute a digest over a subset of their data, e.g. to
+exclude the area where the digest that will be compared against is stored. 
+These types have a source slice and maintain an active range over that slice.
+The digest will be computed only over the active range, rather than the
+entire slice.
+
+```rust
+pub trait DigestData<'a, const L: usize> {
+    fn set_data_client(&'a self, client: &'a dyn ClientData<'a, L>) {}
+    fn add_data(&self, data: LeasableBuffer<'static, u8>) 
+       -> Result<(), (ErrorCode, LeasableBuffer<'static, u8>)>;
+    fn add_mut_data(&self, data: LeasableMutableBuffer<'static, u8>)
+       -> Result<(), (ErrorCode, LeasableMutableBuffer<'static, u8>)>;
+    fn clear_data(&self);
+}
+```
+
+A successful call to `add_data` or `add_mut_data` will add all of the
+data in the active range of the leasable buffer as input to the hash
+function. A successful call is one which returns `Ok(())` and whose
+completion event passes `Ok(())`. If a client needs to compute a hash over several non-contiguous
+regions of a slice, or multiple slices, it can call these methods multiple
+times. 
+
+There may only be one outstanding `add_data`  or `add_mut_data` operation at
+any time. If either `add_data` or `add_mut_data` returns `Ok(())`, then all
+subsequent calls to `add_data` or `add_mut_data` MUST return `Err((ErrorCode::BUSY, ...))`
+until a completion callback delivered through `ClientData`.
+
+```rust
+pub trait ClientData<'a, const L: usize> {
+    fn add_data_done(&'a self, result: Result<(), ErrorCode>, data: LeasableBuffer<'static, u8>);
+    fn add_mut_data_done(
+        &'a self,
+        result: Result<(), ErrorCode>,
+        data: LeasableMutableBuffer<'static, u8>,
+    );
+}
+```
+
+The `data` parameters of `add_data_done` and `add_mut_data_done` indicate what
+data was added and what remains to be added to the digest. If either callback
+has a `result` value of `Ok(())`, then the active region of `data` MUST be zero
+length and all of the data in the active region passed through the corresponding
+call MUST have been added to the digest. 
```

Comment:
This does not match my understanding of `LeasableBuffer`, but as we've talked about on the call of June 3rd, the API is currently underspecified and may require some more work. I think this is okay here, and did write you an email already @phil-levis.

Reply:
These semantics are intended only for `Digest`; it's fine for `LeasableBuffer` to behave differently in different APIs. 

The reasoning here relates to where the iteration occurs. If you have a buffer covering 1000 bytes, but can only do 64 bytes at a time, should it be that the callee iterates 16 times, or the caller? I think it should be the callee, as there's potentially many callers be callee. Replicated code is more likely to be buggy and wastes space.

Reply:
The `LeasableBuffer` return type, though, is extremely useful for the `Err` case -- it can tell you what's remaining. If you were to receive an error code which suggests you can resume, you can pass the buffer back in.